cmaes.lua

```diff
@@ -0,0 +1,273 @@
+   local ftarget = state.ftarget -- stop if fitness < ftarget
+   local maxEval = tonumber(state.maxEval) or 1e3*N^2
+   local objfunc = opfunc
+   local verb_disp = state.verb_disp -- display step size
+   local min_iterations = state.min_iterations or 1
+
+   local lambda = state.popsize -- population size, offspring number
+   -- Strategy parameter setting: Selection  
+   if state.popsize == nil then
+     lambda = 4 + math.floor(3 * math.log(N))
+   end
+
+   local mu = lambda / 2  -- number of parents/points for recombination
+   local weights = torch.range(0,mu-1):apply(function(i) 
+      return math.log(mu+0.5) - math.log(i+1)  end) -- recombination weights
+    weights:div(weights:sum())  -- normalize recombination weights array
+    local mueff = weights:sum()^2 / torch.pow(weights,2):sum()  -- variance-effectiveness of sum w_i x_i
+
+    -- Strategy parameter setting: Adaptation
+    local cc = (4 + mueff/N) / (N+4 + 2 * mueff/N)  -- time constant for cumulation for C
+    local cs = (mueff + 2) / (N + mueff + 5)  -- t-const for cumulation for sigma control
+    local c1 = 2 / ((N + 1.3)^2 + mueff)  -- learning rate for rank-one update of C
+    local cmu = math.min(1 - c1, 2 * (mueff - 2 + 1/mueff) / ((N + 2)^2 + mueff))  -- and for rank-mu update
+    local damps = 2 * mueff/lambda + 0.3 + cs  -- damping for sigma, usually close to 1
+
+    -- Initialize dynamic (internal) state variables 
+    local pc = torch.Tensor(N):zero() -- evolution paths for C
+    local ps = torch.Tensor(N):zero() -- evolution paths for sigma
+    local B = torch.eye(N)   -- B defines the coordinate system 
+    local D = torch.Tensor(N):fill(1)  -- diagonal D defines the scaling
+    local C = torch.eye(N)   -- covariance matrix 
+    local invsqrtC = torch.eye(N)  -- C^-1/2 
+    local eigeneval = 0      -- tracking the update of B and D
+    local counteval = 0
+    local f_hist = {opfunc(x)}  -- for bookkeeping output and termination
+    local best = BestSolution.new(nil,nil,counteval)
+    local iteration = 0 -- iteration of the optimize loop
+
+
+    local function ask()
+      --[[return a list of lambda candidate solutions according to 
+	    m + sig * Normal(0,C) = m + sig * B * D * Normal(0,I)
+	    --]]
+      -- Eigendecomposition: first update B, D and invsqrtC from C
+      -- postpone in case to achieve O(N^2)
+      if counteval - eigeneval > lambda/(c1+cmu)/C:size(1)/10 then
+        eigeneval = counteval
+        C = torch.triu(C) + torch.triu(C,1):t() -- enforce symmetry
+        D, B = torch.symeig(C,'V') -- eigen decomposition, B==normalized eigenvectors, O(N^3)
+        D = torch.sqrt(D)  -- D contains standard deviations now
+        invsqrtC = B * torch.diag(torch.pow(D,-1)) * B:t()
+      end
+      res = torch.Tensor(lambda,D:size(1))
+      for k=1,lambda do --repeat lambda times
+        z = D:clone()
+        z:apply(function(d) return d * torch.normal(0,1) end)
```

Comment:
It would be faster to do:
```
        z = D:clone():normal(0,1):cmul(D)
```

And if you want additional speed improvements, the tensors can be allocated just once.